3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/AnonIpEnrichment.scala

```diff
@@ -118,23 +167,60 @@ case class AnonIpEnrichment(
    * anonymizeIp("94.15.223.151", Three)
    * => "94.x.x.x"
    *
+   *
+   * For IPv6 either the form defined in RFC 2732
+   *  or the literal IPv6 address format defined in RFC 2373 is accepted
+   *
    * TODO: potentially update this to return
    * a Validation error or a null if the IP
    * address is somehow invalid or incomplete.
    *
    * @param ip The IP address to anonymize
    * @return the anonymized IP address
    */
-  import AnonOctets._
-  def anonymizeIp(ip: String): String =
-    Option(ip)
-      .map(
-        _.split("\\.").zipWithIndex
-          .map {
-            case (q, i) => {
-              if (octets.id >= All.id - i) "x" else q
-            }
-          }
-          .mkString("."))
-      .orNull
+  val IPv4MappedAddressPrefix = "::FFFF:"
+  val MaskChar                = "x"
+
+  def anonymizeIp(ipOrNull: String): String =
+    Option(ipOrNull).map { ip =>
+      Try(InetAddress.getByName(ip))
+        .map {
+          case _: Inet4Address => anonymizeIpV4(ip)
+          case _: Inet6Address => anonymizeIpV6(ip)
+        }
+        .getOrElse(tryAnonymizingInvalidIp(ip))
+    }.orNull
+
+  private def anonymizeIpV4(ip: String): String = {
+    def masked(ipv4: String) = {
+      val splitted = ipv4.split("\\.")
+      splitted
+        .take(AnonIPv4Octets.All.id - ipv4Octets.id)
+        .toList
+        .padTo(splitted.size, MaskChar)
+        .mkString(".")
+    }
+
+    if (ip.startsWith(IPv4MappedAddressPrefix))
+      IPv4MappedAddressPrefix + masked(ip.replace(IPv4MappedAddressPrefix, ""))
+    else masked(ip)
+  }
+
+  private def anonymizeIpV6(ip: String): String =
+    InetAddress
+      .getByName(ip)
+      .getHostAddress
+      .split(":")
+      .take(AnonIPv6Segments.All.id - ipv6Segments.id)
+      .toList
+      .padTo(AnonIPv6Segments.All.id, MaskChar)
+      .mkString(":")
+
+  /**
+   * Mainly to not brake code that already exists, i.e. broken IP like this: "777.2.23"
+   * */
+  private def tryAnonymizingInvalidIp(ip: String): String =
```

Comment:
What if we have an invalid IPv6?

Reply:
It's not possible to have invalid IPv6 I think. The basic cases (i.e. null, just string) are handled by IPv4 (you can not know if this was meat to be IPv6). When you have a missing segment in IPv6 the `InetAddress.getByName` deals with it - that's why I'm using it. 
Ok, I just came up with one more broken case, like 123z4::2653 (string that contains not hex number char in it). I'll add test and some fix for it.

Reply:
I've seen somewhere on Zendesk that some customers receive IPv6s with `.` instead of `:`. Would such IPs succeed ?

Reply:
`2605.2700.0.3..4713.93e3` will produce `x.x.x.x.x.x.x` for `Six`
`2605.2700.0.3..4713.93e3` will produce `2605.2700.x.x.x.x.x.x` for `Four` 
I don't know if we could anything reasonable about it.

Reply:
I would say that the solution is not to have a workaround for such cases here but rather to fix the upstream to not receive IPv6 with such formats.